prow/cmd/moonraker/main.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"net/http"
+	"os"
+	"os/exec"
+	"strconv"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+
+	"k8s.io/test-infra/pkg/flagutil"
+	"k8s.io/test-infra/prow/config"
+	prowflagutil "k8s.io/test-infra/prow/flagutil"
+	configflagutil "k8s.io/test-infra/prow/flagutil/config"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/metrics"
+	"k8s.io/test-infra/prow/moonraker"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type options struct {
+	client         prowflagutil.KubernetesOptions
+	github         prowflagutil.GitHubOptions
+	port           int
+	cookiefilePath string
+
+	config configflagutil.ConfigOptions
+
+	dryRun                 bool
+	gracePeriod            time.Duration
+	instrumentationOptions prowflagutil.InstrumentationOptions
+}
+
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+	fs.IntVar(&o.port, "port", 8080, "HTTP port.")
+	fs.DurationVar(&o.gracePeriod, "grace-period", 180*time.Second, "On shutdown, try to handle remaining events for the specified duration. ")
```

Comment:
Hook uses 180 seconds also (which was probably the example copied into Sub, which I must have in turn used in Gangway). I see according to https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase that the grace period used by K8s is just 30s. Will update to 25 seconds.

I can update Hook, Sub, and Gangway to use 25 seconds in a followup.

Reply:
Perhaps you already planned to do this, but a comment relating the figure to the k8s grace period might help future maintainers.

But curious, is it not possible to discover the k8s grace period setting programmatically? It would seem a better way to manage this in general...

Reply:
> Perhaps you already planned to do this, but a comment relating the figure to the k8s grace period might help future maintainers.

Will do.

> But curious, is it not possible to discover the k8s grace period setting programmatically? It would seem a better way to manage this in general...

I think it is possible but it is not something that we do in our other services. The point here is to use a default for our service that is not larger than the k8s *default* grace period. I don't think it's worth the trouble of figuring out programmatically what the k8s default grace period is.